### DISCUSSION TITLE
Display the terminal output in the web page?

### DISCUSSION THREAD

**Discussion by MadeInPierre** [zauberzeug/nicegui#725] [Q&A] [answered]
### Question

Hi everyone!

I would like to add an element in my web page that shows the terminal output (monospace font, colored).

I'm just starting with this project and would appreciate any guidance. Is there a way to add such an element?

Thanks a lot!

**Comment by rodja** [comment, marked as answer]
Thanks for this interesting question @MadeInPierre. Would you like to see the output of the actual script which is currently running? Or the output of other commands? The latter is demoed in our [script executor](https://github.com/zauberzeug/nicegui/blob/main/examples/script_executor/main.py) example. What kind of coloring do you expect. Have you already colored output or what would the coloring rules be?

**Reply by MadeInPierre** [reply]
Thank you @rodja for your interest and quick reply! I'm in the middle of a [personal project](https://github.com/MadeInPierre/finalynx) that currently outputs a colorful tree in the terminal (using the [`rich`](https://github.com/Textualize/rich) library to manage the tree structure and colors which builds on [ANSI codes](https://stackoverflow.com/questions/4842424/list-of-ansi-color-escape-sequences)):

<p align="center">
  <img src="https://raw.githubusercontent.com/MadeInPierre/finalynx/main/docs/_static/screenshot.png" width="300" />
</p>

Now, I'd like to extend this console view with an interactive web dashboard using NiceGUI (awesome project!) by displaying the same tree in the web. I've tried to use NiceGUI's `Tree` element (_please excuse the poor design, this is 100% work in progress_ :sweat_smile:). However, I couldn't find how to customize the tree with colors, badges, interactive elements, ...

![image](https://user-images.githubusercontent.com/12670280/230294244-e3f5f3cc-77ef-4525-87a2-88475df32709.png)

Another approach would be to print the console output in the web dashboard too. I can either capture `rich`'s stdout with raw ANSI codes or export an HTML file with the output colors.

Any guidance on how to transpose this colorful tree in the web using NiceGUI would be greatly appreciated, thanks!

**Reply by falkoschindler** [reply]
Very interesting application! 😀
You say you could export an HTML file. Maybe you just need to plug the HTML content in an `ui.html` element?
Alternatively I could imagine implementing scoped slots for `ui.tree` like it has been done for `ui.table` in this demo: https://nicegui.io/documentation/table#table_with_expandable_rows But this is probably more work than simply using an HTML export.

**Reply by rodja** [reply]
Thats great. `rich` has html and svg export capabilities: https://rich.readthedocs.io/en/stable/console.html#exporting
Should work out of the box with `ui.html`.

**Reply by MadeInPierre** [reply]
Thanks a lot for your guidance! It was simpler than expected... looks great (I'll look into adjusting the font size).

> Alternatively I could imagine implementing scoped slots for ui.tree like it has been done for ui.table in this demo

I hope it's not too much to ask, but I'd be super interested in an equivalent example for Tree! My final goal is to keep the NiceGUI's Tree object and be able to click on each node to update all graphs accordingly, but I'm still starting up with web development.

**Reply by falkoschindler** [reply]
@MadeInPierre Good to hear that it works already!

Here is a demo showing scoped slots with `ui.tree`:

https://github.com/zauberzeug/nicegui/blob/38d24b0ec5fa134dc1180322070ad9a5bccb9b23/website/more_documentation/tree_documentation.py#L13-L35

It will be added [to the documentation](https://nicegui.io/documentation/tree) as soon as we push a new update.

**Reply by MadeInPierre** [reply]
Wow, I wouldn't have hoped for such a quick and useful help. Thank you guys, and good luck for this awesome project!
